java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DatabaseCommands.java

```diff
@@ -0,0 +1,160 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.java.lsp.server.db;
+
+import java.net.URL;
+import java.sql.DatabaseMetaData;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import org.eclipse.lsp4j.MessageParams;
+import org.eclipse.lsp4j.MessageType;
+import org.netbeans.api.db.explorer.ConnectionManager;
+import org.netbeans.api.db.explorer.DatabaseConnection;
+import org.netbeans.api.db.explorer.DatabaseException;
+import org.netbeans.api.db.explorer.JDBCDriver;
+import org.netbeans.api.db.explorer.JDBCDriverManager;
+import org.netbeans.modules.java.lsp.server.protocol.NbCodeLanguageClient;
+import org.netbeans.modules.java.lsp.server.protocol.QuickPickItem;
+import org.netbeans.modules.java.lsp.server.protocol.ShowInputBoxParams;
+import org.netbeans.modules.java.lsp.server.protocol.ShowQuickPickParams;
+import org.openide.filesystems.FileObject;
+import org.openide.filesystems.URLMapper;
+import org.openide.util.NbBundle;
+
+/**
+ *
+ * @author Jan Horvath
+ */
+@NbBundle.Messages({
+    "MSG_EnterDbUrl=Enter DB URL",
+    "MSG_EnterUsername=Enter Username",
+    "MSG_EnterPassword=Enter Password",
+    "MSG_SelectDriver=Select db driver",
+    "MSG_DriverNotFound=Driver not found",
+    "MSG_ConnectionAdded=Connection added",
+    "MSG_ConnectionFailed=Connection failed",
+    "MSG_SelectSchema=Select Database Schema"
+})
+public class DatabaseCommands {
+
+    public static CompletableFuture<Object> addDbConnection(NbCodeLanguageClient client) {
+        JDBCDriver[] drivers = JDBCDriverManager.getDefault().getDrivers();
+        List<QuickPickItem> items = new ArrayList<>();
+        for (int i = 0; i < drivers.length; i++) {
+            URL[] jars = drivers[i].getURLs();
+            if (jars != null && jars.length > 0) {
+                FileObject jarFO = URLMapper.findFileObject(jars[0]);
+                if (jarFO != null && jarFO.isValid()) {
+                    items.add(
+                            new QuickPickItem(drivers[i].getName(), null, drivers[i].getDisplayName() + " (" + drivers[i].getClassName() + ")", false, i) // NOI18N
+                    );
+                }
+            }
+        }
+
+        return client.showQuickPick(new ShowQuickPickParams(Bundle.MSG_SelectDriver(), false, items))
+                .thenApply(selectedItems -> {
+                    if (selectedItems == null) {
+                        return null;
+                    }
+                    if (!selectedItems.isEmpty()) {
+                        int i = ((Double) selectedItems.get(0).getUserData()).intValue();
```

Comment:
What am I missing here? Why is this a `Double` and not an `Integer` as declared in line 69? If this is a lossy conversion, I would cast to `Number`, shielding you from implementation specific brokenness.

Reply:
The trouble is that the `QuickPickItem` structure is mashalled to the LSP connection stream and goes to the client and back. And because the client is (yuck!) javascript, it represents integer as `number` which is unmarshalled as double... At this moment I do not know if the culprit is eclipse lsp4j, it is a feature by design (who the knows what type 2 actually is in a JSON document) or the javascript existence itself. If it is Eclipse stuff, we could eventually file a bug against the library.
To improve possible future compatibility, I'd put there 
`((Number) selectedItems.get(0).getUserData()).intValue();`
so if ever numbers w/o floating part deserialize to Integers or Longs the code survives.

Reply:
That makes sense. Thanks for the pointer.